[POS][Custom payment UI] – Payment cancelation

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13006
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR implements payment flow cancellation in POS, by providing a custom coroutine scope to the payment controller. That coroutine scope is used to cancel the card reader payment [action](https://github.com/woocommerce/woocommerce-android/blob/0d7f4c53ef09024b6699e2b8dfd2a2eec34e1536/libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/actions/CollectPaymentAction.kt#L42).

⚠️ Don't merge until the base branch is not merged into trunk.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Go to Woo POS (Menu more > Point of Sale)
2. Ensure reader is connected
3. Add items to cart and click "Checkout"
4. Wait until the order draft is connected and reader is ready to collect card payment (LoadingData -> ExternalReaderCollectPaymentState)
5. Click system "back" button
6. Observe that reader cancels payment action (M2 is playing sound, WisePad clears screen)
7. Verify it's possible to collect another payment now (previously it would not be possible because payment action was not canceled)

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Verify that the payment action is canceled on back press in checkout and it's possible to collect another payment.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
* I verified that reader payment action is canceled and consecutive payment actions are possible.
* Verified that paymentScope.cancel() doesn't cancel viewModelScope
* Verified that `WooPosTotalsViewModle.viewModelScope.cancel()` cancels `paymentScope`
* Verified that when `paymentScope.cancel()` the `CollectPaymentAction` is canceled

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/b2b53782-f744-4ce4-8853-c18bcef7c9d6

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Restrict cancelable states by common interface

To be able to filter them using when expression.

samiuelson committed:
Attempt to cancel a payment on back press

samiuelson committed:
Revert "Restrict cancelable states by common interface"

This reverts commit a02047d10024b546a79410a8a2445422514cac5b.

samiuelson committed:
Use custom, cancelable payment scope

samiuelson committed:
Update paymentScope to inherit from viewModel scope

samiuelson committed:
Remove unused imports

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -73,13 +77,15 @@ class WooPosTotalsViewModel @Inject constructor(
     private var cardReaderPaymentController: CardReaderPaymentController? = null
 
     private fun createCardReaderPaymentController(orderId: Long) {
+        paymentScope = CoroutineScope(SupervisorJob(viewModelScope.coroutineContext[Job]))
```
💡 `paymentScope` characteristics:
* `paymentScope` gets canceled automatically when `viewModelScope` is canceled, because the new `SupervisorJob` here inherits from `viewModelScope.coroutineContext[Job]` (which is internal viewModel's SupervisorJob, that's canceled when VM is destroyed)
* when `paymentScope` is canceled the `viewModelScope` is not canceled automatically (thanks to `SupervisorJob`)
* Default dispatchers are inherited from `viewModelScope` (viewModelScope.coroutineContext; Dispatchers.Main.immediate)

kidinov replied:
@samiuelson 

Please correct if my understanding is not correct. We need to have a scope that gets canceled in 2 cases:
* When the object that keeps `CardReaderPaymentController` is getting destroyed. In our cases, those are 2 VMs
* When we manually want to cancel it, e.g. when back clicked in the POS case

I am thinking, as we have `onStart` and `onCleared` methods, which we must call anyway, would it make sense to create scope inside of `CardReaderPaymentController:onStart` and cancel in `onCleared` (btw I'd call it `onStop`)? 

There are a couple of advantages I see:
* No need to construct `CoroutineScope(SupervisorJob(viewModelScope.coroutineContext[Job]))` scope, which is not easy to follow, to be honest
* We don't need to pass scope at all there and it's handled by the controller
* No need to handle the scope on the caller side too

It has one disadvantage too - if a client misses the `onCleared` call then the payment won't be canceled. Having said that now if the passed scoped not canceled at the right moment the same happens

Wdyt?

samiuelson replied:
I tried to avoid modifying the code in payment controller as much as possible, but in this case I think it's a good point 👍 

I modified controller to create and manage the coroutine scope internally – [e499d89](https://github.com/woocommerce/woocommerce-android/pull/13037/commits/e499d895a884fd4819fc49f180ca815a7c042ad5)

Let me know what you think, @kidinov!

samiuelson committed:
Remove redundant `onClear` implementation

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -207,10 +215,6 @@ class WooPosTotalsViewModel @Inject constructor(
     }
 
-    override fun onCleared() {
-        cardReaderPaymentController?.onCleared()
```
💡 `cardReaderPaymentController?.onCleared()` should be called when the `paymentScope` is cancelled

kidinov replied:
Why? My understanding is that the "onClose" callback of the "flow" will be called, but this method won't

samiuelson replied:
Fixed – e499d895a884fd4819fc49f180ca815a7c042ad5

kidinov reviewed (commented):
@samiuelson I left a few comments, please take a look!

No related to the PR:`CardReaderPaymentViewModel::reFetchOrder` not invoked anymore. Is this ok?

samiuelson committed:
Manage payment coroutine scope internally in controller

samiuelson committed:
Clean up code

samiuelson committed:
Remove unused code

I've checked the history, and `CardReaderPaymentViewModel::reFetchOrder` has never been invoked outside of the class. When extracting payment flow to the controller, I missed the `@VisibleForTesting` annotation and treated it as a public function when it was supposed to be treated as `private`. I'm removing this method from VM – it's supposed to be invoked internally in `CardReaderPaymentController`.

samiuelson commented:
> No related to the PR:`CardReaderPaymentViewModel::reFetchOrder` not invoked anymore. Is this ok?

That's a good catch, @kidinov. I've checked the history, and `CardReaderPaymentViewModel::reFetchOrder` has never been invoked outside of the class. When extracting payment flow to the controller, I missed the `@VisibleForTesting` annotation and treated it as a public function when it was supposed to be treated as `private`. I'm removing this method from VM – it's supposed to be invoked internally in `CardReaderPaymentController` – ad8397301d46b5286cb6db8dee0316ddb9664cfb

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentControllerTest.kt:
```diff
@@ -301,6 +303,7 @@ class CardReaderPaymentControllerTest : BaseUnitTest() {
         testBlocking {
             whenever(cardReaderManager.displayBluetoothCardReaderMessages).thenAnswer {
                 flow {
+                    delay(1) // Ensure CardReaderDisplayMessage is emitted after flow is at CollectingPayment state
```
I'd really avoid having delays in the tests -> may lead to flakiness. Is there a way to wait for the `CollectingPayment` event and then emit it?

samiuelson replied:
👍 Removing `delay()` from all the tests in controller – 8f715b4

kidinov reviewed (approved):
LGTM!

I left one comment regarding the test, please take a look

samiuelson committed:
Remove delay(1) from tests
